import unittest
import sys
from icecream import ic
from COMPARISONS import COMPARISONS



class TestCOMPARISONS(unittest.TestCase):
    """
    rather weak test for a week objects set...
    """

    pair = 'ena_cl::mixs_v6'
    comparisonStats = {'ena_cl::mixs_v6': {'by_package': {
        'COMPARE-ECDC-EFSA pilot food-associated reporting standard::Agriculture': {'pc_left_of_right': 0.01,
                                                                                    'length_left': 7,
                                                                                    'length_right': 98,
                                                                                    'length_union': 104,
                                                                                    'length_clean_intersection': 1,
                                                                                    'clean_intersection_set': {
                                                                                        'collection_date'},
                                                                                    'clean_left_diff_set': {'isolate',
                                                                                                            'serotype',
                                                                                                            'isolation_source',
                                                                                                            'geographic_location_(region_and_locality)',
                                                                                                            'geographic_location_(country_and_or_sea)',
                                                                                                            'collected_by'},
                                                                                    'clean_right_diff_set': {
                                                                                        'assembly_software',
                                                                                        'annotation',
                                                                                        'trna_extraction_software',
                                                                                        'library_screening_strategy',
                                                                                        'single_cell_lysis_appr',
                                                                                        'sample_material_processing',
                                                                                        'taxonomic_identity_marker',
                                                                                        'library_vector',
                                                                                        'relationship_to_oxygen',
                                                                                        'extrachromosomal_elements',
                                                                                        'pcr_conditions',
                                                                                        'local_environmental_context',
                                                                                        'geographic_location_(country_and_or_sea,region)',
                                                                                        'target_gene',
                                                                                        'chimera_check_software',
                                                                                        'broad_scale_environmental_context',
                                                                                        'sequencing_method',
                                                                                        'host_prediction_estimated_accuracy',
                                                                                        'library_layout',
                                                                                        'completeness_approach',
                                                                                        'viral_identification_software',
                                                                                        'mag_coverage_software',
                                                                                        'otu_classification_approach',
                                                                                        'binning_software',
                                                                                        'source_of_uvigs',
                                                                                        'completeness_score',
                                                                                        'x_16s_recover', 'sample_name',
                                                                                        'trophic_level',
                                                                                        'library_reads_sequenced',
                                                                                        'pcr_primers', 'otu_database',
                                                                                        'reference_for_biomaterial',
                                                                                        'relevant_electronic_resources',
                                                                                        'ploidy',
                                                                                        'contamination_screening_parameters',
                                                                                        'binning_parameters',
                                                                                        'reference_database(s)',
                                                                                        'relevant_standard_operating_procedures',
                                                                                        'sample_volume_or_weight_for_dna_extraction',
                                                                                        'estimated_size',
                                                                                        'samp_collec_device',
                                                                                        'library_size',
                                                                                        'isolation_and_growth_condition',
                                                                                        'target_subfragment',
                                                                                        'number_of_standard_trnas_extracted',
                                                                                        'depth',
                                                                                        'completeness_software',
                                                                                        'nucleic_acid_extraction',
                                                                                        'host_specificity_or_range',
                                                                                        'size_fraction_selected',
                                                                                        'number_of_contigs',
                                                                                        'virus_enrichment_approach',
                                                                                        'experimental_factor',
                                                                                        'sorting_technology',
                                                                                        'temperature',
                                                                                        'host_prediction_approach',
                                                                                        'elevation',
                                                                                        'otu_sequence_comparison_approach',
                                                                                        'multiplex_identifiers',
                                                                                        'taxonomic_classification',
                                                                                        'samp_collec_method',
                                                                                        'adapters',
                                                                                        'number_of_replicons',
                                                                                        'amount_or_size_of_sample_collected',
                                                                                        'reassembly_post_binning',
                                                                                        'observed_biotic_relationship',
                                                                                        'x_16s_recover_software',
                                                                                        'feature_prediction',
                                                                                        'geographic_location_(latitude_and_longitude)',
                                                                                        'wga_amplification_approach',
                                                                                        'sequence_quality_check',
                                                                                        'source_material_identifiers',
                                                                                        'subspecific_genetic_lineage',
                                                                                        'assembly_quality',
                                                                                        'detection_type',
                                                                                        'nucleic_acid_amplification',
                                                                                        'project_name',
                                                                                        'predicted_genome_structure',
                                                                                        'taxonomy_id_of_dna_sample',
                                                                                        'host_disease_status',
                                                                                        'assembly_name',
                                                                                        'known_pathogenicity',
                                                                                        'similarity_search_method',
                                                                                        'negative_control_type',
                                                                                        'predicted_genome_type',
                                                                                        'single_cell_lysis_prot',
                                                                                        'host_scientific_name',
                                                                                        'positive_control_type',
                                                                                        'wga_amplification_kit',
                                                                                        'environmental_medium',
                                                                                        'contamination_score',
                                                                                        'decontamination_software',
                                                                                        'encoded_traits', 'propagation',
                                                                                        'altitude',
                                                                                        'contamination_screening_input'},
                                                                                    'left_diff_set': {'isolate',
                                                                                                      'serotype',
                                                                                                      'geographic location (country and/or sea)',
                                                                                                      'isolation_source',
                                                                                                      'geographic location (region and locality)',
                                                                                                      'collected_by'},
                                                                                    'right_diff_set': {
                                                                                        'OTU classification approach',
                                                                                        'annotation',
                                                                                        'sequencing method',
                                                                                        'host specificity or range',
                                                                                        'relationship to oxygen',
                                                                                        'viral identification software',
                                                                                        'host prediction estimated accuracy',
                                                                                        'single_cell_lysis_appr',
                                                                                        'sorting technology',
                                                                                        'OTU sequence comparison approach',
                                                                                        'detection type',
                                                                                        'trophic level',
                                                                                        'pcr conditions',
                                                                                        'geographic location (country and/or sea,region)',
                                                                                        'multiplex identifiers',
                                                                                        'sample volume or weight for DNA extraction',
                                                                                        'broad-scale environmental context',
                                                                                        'sequence quality check',
                                                                                        'predicted genome type',
                                                                                        'library layout',
                                                                                        'chimera check software',
                                                                                        'reassembly post binning',
                                                                                        'completeness approach',
                                                                                        'encoded traits',
                                                                                        'binning parameters',
                                                                                        'amount or size of sample collected',
                                                                                        'x_16s_recover',
                                                                                        'MAG coverage software',
                                                                                        'WGA amplification approach',
                                                                                        'geographic location (latitude and longitude)',
                                                                                        'number of contigs',
                                                                                        'feature prediction',
                                                                                        'project name', 'sample name',
                                                                                        'completeness score',
                                                                                        'assembly software',
                                                                                        'observed biotic relationship',
                                                                                        'nucleic acid extraction',
                                                                                        'decontamination software',
                                                                                        'library vector',
                                                                                        'contamination score',
                                                                                        'binning software',
                                                                                        'similarity search method',
                                                                                        'relevant standard operating procedures',
                                                                                        'OTU database', 'ploidy',
                                                                                        'estimated size',
                                                                                        'isolation and growth condition',
                                                                                        'nucleic acid amplification',
                                                                                        'target gene',
                                                                                        'experimental factor',
                                                                                        'negative control type',
                                                                                        'samp_collec_device',
                                                                                        'reference for biomaterial',
                                                                                        'contamination screening input',
                                                                                        'pcr primers',
                                                                                        'target subfragment',
                                                                                        'virus enrichment approach',
                                                                                        'source of UViGs',
                                                                                        'reference database(s)',
                                                                                        'assembly name',
                                                                                        'host disease status', 'depth',
                                                                                        'sample material processing',
                                                                                        'size fraction selected',
                                                                                        'taxonomic classification',
                                                                                        'assembly quality',
                                                                                        'completeness software',
                                                                                        'subspecific genetic lineage',
                                                                                        'number of standard tRNAs extracted',
                                                                                        'library reads sequenced',
                                                                                        'positive control type',
                                                                                        'temperature',
                                                                                        'known pathogenicity',
                                                                                        'elevation',
                                                                                        'environmental medium',
                                                                                        'taxonomy ID of DNA sample',
                                                                                        'samp_collec_method',
                                                                                        'adapters',
                                                                                        'host scientific name',
                                                                                        'taxonomic identity marker',
                                                                                        'x_16s_recover_software',
                                                                                        'WGA amplification kit',
                                                                                        'predicted genome structure',
                                                                                        'tRNA extraction software',
                                                                                        'relevant electronic resources',
                                                                                        'number of replicons',
                                                                                        'contamination screening parameters',
                                                                                        'extrachromosomal elements',
                                                                                        'library screening strategy',
                                                                                        'single_cell_lysis_prot',
                                                                                        'propagation', 'altitude',
                                                                                        'local environmental context',
                                                                                        'host prediction approach',
                                                                                        'source material identifiers',
                                                                                        'library size'},
                                                                                    'intersection_set': {
                                                                                        'collection date'}},
        'COMPARE-ECDC-EFSA pilot food-associated reporting standard::AgricultureMIGSBacteria': {
            'pc_left_of_right': 0.01, 'length_left': 7, 'length_right': 98, 'length_union': 104,
            'length_clean_intersection': 1, 'clean_intersection_set': {'collection_date'},
            'clean_left_diff_set': {'isolate', 'serotype', 'isolation_source',
                                    'geographic_location_(region_and_locality)',
                                    'geographic_location_(country_and_or_sea)', 'collected_by'},
            'clean_right_diff_set': {'assembly_software', 'annotation', 'trna_extraction_software',
                                     'library_screening_strategy', 'single_cell_lysis_appr',
                                     'sample_material_processing', 'taxonomic_identity_marker', 'library_vector',
                                     'relationship_to_oxygen', 'extrachromosomal_elements', 'pcr_conditions',
                                     'local_environmental_context', 'geographic_location_(country_and_or_sea,region)',
                                     'target_gene', 'chimera_check_software', 'broad_scale_environmental_context',
                                     'sequencing_method', 'host_prediction_estimated_accuracy', 'library_layout',
                                     'completeness_approach', 'viral_identification_software', 'mag_coverage_software',
                                     'otu_classification_approach', 'binning_software', 'source_of_uvigs',
                                     'completeness_score', 'x_16s_recover', 'sample_name', 'trophic_level',
                                     'library_reads_sequenced', 'pcr_primers', 'otu_database',
                                     'reference_for_biomaterial', 'relevant_electronic_resources', 'ploidy',
                                     'contamination_screening_parameters', 'binning_parameters',
                                     'reference_database(s)', 'relevant_standard_operating_procedures',
                                     'sample_volume_or_weight_for_dna_extraction', 'estimated_size',
                                     'samp_collec_device', 'library_size', 'isolation_and_growth_condition',
                                     'target_subfragment', 'number_of_standard_trnas_extracted', 'depth',
                                     'completeness_software', 'nucleic_acid_extraction', 'host_specificity_or_range',
                                     'size_fraction_selected', 'number_of_contigs', 'virus_enrichment_approach',
                                     'experimental_factor', 'sorting_technology', 'temperature',
                                     'host_prediction_approach', 'elevation', 'otu_sequence_comparison_approach',
                                     'multiplex_identifiers', 'taxonomic_classification', 'samp_collec_method',
                                     'adapters', 'number_of_replicons', 'amount_or_size_of_sample_collected',
                                     'reassembly_post_binning', 'observed_biotic_relationship',
                                     'x_16s_recover_software', 'feature_prediction',
                                     'geographic_location_(latitude_and_longitude)', 'wga_amplification_approach',
                                     'sequence_quality_check', 'source_material_identifiers',
                                     'subspecific_genetic_lineage', 'assembly_quality', 'detection_type',
                                     'nucleic_acid_amplification', 'project_name', 'predicted_genome_structure',
                                     'taxonomy_id_of_dna_sample', 'host_disease_status', 'assembly_name',
                                     'known_pathogenicity', 'similarity_search_method', 'negative_control_type',
                                     'predicted_genome_type', 'single_cell_lysis_prot', 'host_scientific_name',
                                     'positive_control_type', 'wga_amplification_kit', 'environmental_medium',
                                     'contamination_score', 'decontamination_software', 'encoded_traits', 'propagation',
                                     'altitude', 'contamination_screening_input'},
            'left_diff_set': {'isolate', 'serotype', 'geographic location (country and/or sea)', 'isolation_source',
                              'geographic location (region and locality)', 'collected_by'},
            'right_diff_set': {'OTU classification approach', 'annotation', 'sequencing method',
                               'host specificity or range', 'relationship to oxygen', 'viral identification software',
                               'host prediction estimated accuracy', 'single_cell_lysis_appr', 'sorting technology',
                               'OTU sequence comparison approach', 'detection type', 'trophic level', 'pcr conditions',
                               'geographic location (country and/or sea,region)', 'multiplex identifiers',
                               'sample volume or weight for DNA extraction', 'broad-scale environmental context',
                               'sequence quality check', 'predicted genome type', 'library layout',
                               'chimera check software', 'reassembly post binning', 'completeness approach',
                               'encoded traits', 'binning parameters', 'amount or size of sample collected',
                               'x_16s_recover', 'MAG coverage software', 'WGA amplification approach',
                               'geographic location (latitude and longitude)', 'number of contigs',
                               'feature prediction', 'project name', 'sample name', 'completeness score',
                               'assembly software', 'observed biotic relationship', 'nucleic acid extraction',
                               'decontamination software', 'library vector', 'contamination score', 'binning software',
                               'similarity search method', 'relevant standard operating procedures', 'OTU database',
                               'ploidy', 'estimated size', 'isolation and growth condition',
                               'nucleic acid amplification', 'target gene', 'experimental factor',
                               'negative control type', 'samp_collec_device', 'reference for biomaterial',
                               'contamination screening input', 'pcr primers', 'target subfragment',
                               'virus enrichment approach', 'source of UViGs', 'reference database(s)', 'assembly name',
                               'host disease status', 'depth', 'sample material processing', 'size fraction selected',
                               'taxonomic classification', 'assembly quality', 'completeness software',
                               'subspecific genetic lineage', 'number of standard tRNAs extracted',
                               'library reads sequenced', 'positive control type', 'temperature', 'known pathogenicity',
                               'elevation', 'environmental medium', 'taxonomy ID of DNA sample', 'samp_collec_method',
                               'adapters', 'host scientific name', 'taxonomic identity marker',
                               'x_16s_recover_software', 'WGA amplification kit', 'predicted genome structure',
                               'tRNA extraction software', 'relevant electronic resources', 'number of replicons',
                               'contamination screening parameters', 'extrachromosomal elements',
                               'library screening strategy', 'single_cell_lysis_prot', 'propagation', 'altitude',
                               'local environmental context', 'host prediction approach', 'source material identifiers',
                               'library size'}, 'intersection_set': {'collection date'}}}}}

    comparison_obj = COMPARISONS(comparisonStats, pair)
    def test_left_source(self):
        self.assertEqual(self.comparison_obj.left_source(),"ena_cl")

    def test_right_source(self):
        self.assertEqual(self.comparison_obj.right_source(), 'mixs_v6')

    def test_put_reorg_df(self):
        df = self.comparison_obj.reorg_df
        ic(self.comparison_obj.put_reorg_df(df))
        self.assertGreater(len(df), 1)

    # def test_get_max_intersection_len(self):
    #     ic(self.comparison_obj.max_intersection_len)
    #     self.assertEqual(self.comparison_obj.max_intersection_len,3)

    def test_ingest(self):
        self.assertEqual(self.comparison_obj.left_source(), "ena_cl")




if __name__ == '__main__':
    unittest.main()

